Added Team, Role and TeamMembership models

xoen opened the pull request:
The `TeamMembership` model keep the relationship between teams, users and
roles.
A user belongs to a team with a given role (e.g. "maintainer" or "member").

A user can be in a team only once and only with a single role:
  * add a user to a team more than once doesn't make any sense
  * add a user to a team with more than one role is confusing and doesn't
    make any sense (e.g. has the user the "member" role or the "maintainer"
    role?)

I also added these membership utility methods:
  * `User#teams` returns the teams (queryset) the user belongs to
  * `Team#users` returns the users (queryset) in the team
  * `Team#users_with_role` returns the users (queryset) in the team with
    the given `role_code`

andyhd commented:
Why Role "code" and not "name"?

xoen commented:
Thanks for the review @andyhd , I'll add a `name` and use [`django.db.TimeStampedModel`](https://github.com/django-extensions/django-extensions/blob/master/django_extensions/db/models.py#L14) as parent class of our models (and see if breaks).

xoen commented:
@andyhd FYI: I added some tests, the `name` field to roles and changed models to inherit from `django_extensions.db.models.TimeStampedModel`, feel free to have another look.

andyhd commented on control_panel_api/tests/test_models.py:
```diff
@@ -0,0 +1,79 @@
+)
+
+
+class FixtureTestCase(TestCase):
```
This might be better named something like `TeamMembershipTestCase`

andyhd commented on control_panel_api/tests/test_models.py:
```diff
@@ -0,0 +1,79 @@
+class FixtureTestCase(TestCase):
+
+    @classmethod
+    def setUpTestData(cls):
```
Should this be called `setUpClass`? https://docs.python.org/3.6/library/unittest.html#setupclass-and-teardownclass

xoen replied:
Bear in mid that `TestCase` here is an instance of `django.test.TestCase` which provides the`setUpTestData` method for this specific use case: https://docs.djangoproject.com/en/1.11/topics/testing/tools/#django.test.TestCase.setUpTestData

andyhd replied:
👍

andyhd commented on control_panel_api/tests/test_models.py:
```diff
@@ -0,0 +1,79 @@
+            team=cls.team_other,
+            user=cls.user_other,
+            role=cls.role_member,
+        )
```
Should these fixtures be set up with `model_mommy`?

xoen replied:
To be honest I don't see what's the benefit of using it in this case when using the django ORM methods are called in the same way and they're standard.

andyhd replied:
Fair enough - I just thought you were using it already

andyhd commented on control_panel_api/tests/test_models.py:
```diff
@@ -0,0 +1,79 @@
+        alices_teams = self.user_alice.teams()
+
+        assert self.team_justice in alices_teams
+        assert self.team_other not in alices_teams
```
When using `unittest`, these assertions should really be `self.assertIn` and `self.assertNotIn` (https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertIn)

xoen replied:
I'll give it a go 👍

andyhd commented on control_panel_api/tests/test_models.py:
```diff
@@ -0,0 +1,79 @@
+)
+
+
+class FixtureTestCase(TestCase):
```
These tests should all be methods on a single class, rather than subclasses of `FixtureTestCase`.

xoen replied:
Alright, moving everything in a `TestCase` as these tests are all related to the membership 👍

andyhd commented on control_panel_api/tests/test_models.py:
```diff
@@ -0,0 +1,79 @@
+            team=cls.team_other,
+            user=cls.user_other,
+            role=cls.role_member,
+        )
```
These `TeamMembership` objects should not be in a global setup - these are setting up the context for individual tests and should be created in the tests methods that test them. Eg:
```python
def test_alice_is_in_team_justice(self):
    TeamMembership.objects.create(user=alice, team=team_justice, role=maintainer)
    self.assertIn(team_justice, alice.teams())
    self.assertNotIn(team_other, alice.teams())
```

The first set of fixtures (alice, bob, other, and the teams) are like the "Given some users and teams exist" context.
The second set are more like "When Alice is maintainer of Team Justice".

xoen commented:
Rebased off `api` branch and removed unnecessary `("team", "user", "role")` unique constraint, as we already have a unique constraint on `("team", "user")`.

andyhd commented on control_panel_api/tests/test_models.py:
```diff
@@ -0,0 +1,98 @@
+class Membership(FixtureTestCase):
+
+    def test_user_can_be_added_to_team_only_once(self):
+        raised_integrity_error = False
```
This is reimplementing what `unittest` already provides in `self.assertRaises`: https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises

xoen replied:
As I said, these tests are using the `django.test.TestCase` class which is using atomic transactions (to clean the DB, etc...) which means exceptions would be caught and then `assertRaises` would not work: https://stackoverflow.com/a/24589930/455642

andyhd replied:
That SO answer suggests you use `TransactionTestCase` - that seems like a good fit here

andyhd commented on control_panel_api/tests/test_models.py:
```diff
@@ -0,0 +1,89 @@
+            team=cls.team_justice,
+            user=cls.user_alice,
+            role=cls.role_maintainer,
+        )
```
This is a bit verbose - maybe in this test module there should be a function `add_user_to_team(user, team, role="Member")` or maybe a method `User.add_to_team(team, role)` or `Team.add_user(user, role)` or similar?

xoen replied:
Yes, we may want to add some methods like that in the future.

andyhd commented on control_panel_api/tests/test_models.py:
```diff
@@ -0,0 +1,89 @@
+        except IntegrityError:
+            raised_integrity_error = True
+
+        assert raised_integrity_error
```
This last test could be separated to allow using `assertRaises`, eg:
```python
class UniqueMembershipTest(TransactionTestCase):
    ...
    def test_user_can_be_added_to_team_only_once(self):
        with self.assertRaises(IntegrityError):
            team_justice.add_user(alice)
```
I realise this complicates the fixtures setup - you could have module-level variables `users = {}` and use `setUpModule` to create them, then in your tests you can get them with `users['Alice']` or similar.

andyhd replied:
This last `assert` should be `self.assertTrue(raised_integrity_error)`

xoen replied:
Done 👍
